crates/red_knot_python_semantic/src/semantic_index/use_def/symbol_state.rs

```diff
@@ -173,11 +220,12 @@ impl SymbolBindings {
         }
     }
 
-    /// Iterate over currently live bindings for this symbol.
-    pub(super) fn iter(&self) -> BindingIdWithConstraintsIterator {
+    /// Iterate over currently live bindings for this symbol, in reverse order.
+    pub(super) fn iter_rev(&self) -> BindingIdWithConstraintsIterator {
```

Comment:
I don't fully crasp how this works yet, so I might be very wrong here. 

I understand that we'll now iterate from the back when ever looking up a binding. Does that mean that checking now becomes `O(n^2)` if we have something like

```python
x = 1
x += 1 # (iterates over x = 1)
x += 1 # (iterates over x +1, x=1)
x += 1 # (iterates over x + 1, x + 1, x = 1)
... 
```

Reply:
> Does that mean that checking now becomes `O(n^2)` if we have something like

I don't think so. We would only ever iterate over one *live* binding in this example. Rewriting using `x = x + 1` to distinguish stores and reads:
```py
x = 1   # binding 1
x = x + 1  # binding 2; read of x only has one visible binding: 1
x = x + 2  # binding 3; read of x only has one visible binding: 2
```

In general, we never iterate over more bindings than we did before on this branch (there is more work to be done in each iteration though). We only changed the order of the iteration. But we might iterate over *fewer* bindings due to short circuiting in an example like this:
```py
x = 1

if True:
    x = 2

x
```
Previously, we would iterate over both bindings of `x`. Now, while iterating in reverse, we can stop at the `x = 2` binding, because nothing else is visible "beyond" this binding.